src/Build/Instance/ProjectInstance.cs

```diff
@@ -2710,5 +2771,30 @@ private void CreatePropertiesSnapshot(Evaluation.Project.Data data, bool isImmut
                 _properties.Set(instance);
             }
         }
+
+        /// <summary>
+        /// Tracks this property read if it's from an environment variable or uninitialized.
+        /// </summary>
+        /// <param name="name">The name of the property being read.</param>
+        private void TrackPropertyRead(string name, bool propertyMissing)
```

Comment:
I think issuing an event for each of these would be better than storing them in collections:
- There is a lot of duplication in this data, most evaluations will have the same environment reads. This will repetition will become non-trivial when you have 10K evaluations. On the other hand, via an event, the receiver (buildxl, qb, binary log viewer) can choose to deduplicate between multiple project evaluations
- The binary log viewer can only get information via log events, which makes the log events the only way to transport this information. @KirillOsenkov

Implementation wise you could reuse the mechanism that gets used by logging each property reassignment: [https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/Evaluator.cs#L1400-L1422](https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/Evaluator.cs#L1400-L1422) (@rainersigwald for opinion)

Reply:
I'm not sure how I would propagate events from the ProjectInstance, through the GraphBuilder, up to the ProjectGraph that needs it.

Reply:
I think it would require moving the logic to the Evaluator, which already logs a related event, that of property reassignments: https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/Evaluator.cs#L1400-L1422.
Then either Bxl/qb or ProjectGraph via some flag would hook up a logger that picks up the event from all the project evaluations and dedupes it.